dlib/matrix/matrix_fft.h

```diff
@@ -9,836 +9,135 @@
 #include "../algs.h"
 
 #ifdef DLIB_USE_MKL_FFT
-#include <mkl_dfti.h>
-#endif
-
-// No using FFTW until it becomes thread safe!
-#if 0
-#ifdef DLIB_USE_FFTW
-#include <fftw3.h>
-#endif // DLIB_USE_FFTW
+#include "mkl_fft.h"
+#define impl_fft    mkl_fft
+#define impl_fftr   mkl_fftr
+#define impl_fftri  mkl_fftri
+#else
+#include "kiss_fft.h"
+#define impl_fft    kiss_fft
+#define impl_fftr   kiss_fftr
+#define impl_fftri  kiss_fftri
 #endif
 
 namespace dlib
-{
-
+{     
 // ----------------------------------------------------------------------------------------
-
-    inline bool is_power_of_two (
-        const unsigned long& value
-    )
+    
+    constexpr long fftr_nc_size(long nc)
     {
-        if (value == 0)
-            return true;
-        else
-            return count_bits(value) == 1;
+        return nc == 0 ? 0 : nc/2+1;
     }
-
-// ----------------------------------------------------------------------------------------
-
-    namespace impl
+    
+    constexpr long ifftr_nc_size(long nc)
```

Comment:
Add some docs for what these are about.

Reply:
I agree. All libraries i've seen use ```fftr```. So users doing FFTs will know exactly what ```fftr``` is doing.

Reply:
Yeah.  Me too.  It's just always bothered me that it wasn't captured in the type system :|

But me being bothered doesn't really matter.  What's least confusing to users is always right.

Reply:
In the end, i did capture it in the type system right? I just used operator overloading. So functionality was there at least.

Reply:
I mean that the returned type from `fft(some_real_matrix)` would encode the fact that it holds a packed representation.  So that then `ifftr()` wouldn't be required.  You would just do `ifft()` and it would know it was a packed representation.  It's complicated though because someone is going to do stuff to the complex packed matrix, like transform it somehow and the fact that it's packed would need to be propagated all the way through all that too.  I can think of a bunch of ways to go at that problem in the typesystem. I'm not sure what would actually not just end up irritating users though.  

The simplest thing I can think of is to just have a new complex type like `dlib::packed_complex<T>` or something like that.  Then the output of a real fft would be something like `dlib::matrix<dlib::packed_complex<double>>` and `ifft(dlib::matrix<dlib::packed_complex<double>>)` would know to do the packed version.  But now there is this other complex type which is definitely going to be irritating to users.

Anyway, none of the ways to do it are worth it and probably all are confusing to users.